sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala

```diff
@@ -104,21 +104,35 @@ class UDFRegistration private[sql] (functionRegistry: FunctionRegistry) extends
       val inputTypes = (1 to x).foldRight("Nil")((i, s) => {s"ScalaReflection.schemaFor[A$i].dataType :: $s"})
       println(s"""
         /**
-         * Register a Scala closure of ${x} arguments as user-defined function (UDF).
+         * Registers a Scala closure of ${x} arguments as user-defined function (UDF).
          * @tparam RT return type of UDF.
          * @since 1.3.0
          */
         def register[$typeTags](name: String, func: Function$x[$types]): UserDefinedFunction = {
+          registerUDF(name, func, deterministic = true)
+        }
+
+        /**
+         * Registers a non-deterministic Scala closure of ${x} arguments as user-defined function (UDF).
+         * @tparam RT return type of UDF.
+         * @since 2.3.0
+         */
+        def registerNonDeterministic[$typeTags](name: String, func: Function$x[$types]): UserDefinedFunction = {
+          registerUDF(name, func, deterministic = false)
+        }
+
+        private def registerUDF[$typeTags](name: String, func: Function$x[$types], deterministic: Boolean): UserDefinedFunction = {
```

Comment:
My only concern is we have many public functions with different names that are doing the similar things.